### PR TITLE
chore: Setup the cluster cache to only watch required resources

### LINKED ISSUE #183: Only cache required resources
# Summary

Currently, we are watching all kinds of resources for cluster cache. To avoid unnecessarily increase the k8s server load and the cache size, we should only cache numaflow resources. 

# Use Cases

Minimize the cache size.

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by xdevxy**
Fixes #183 

### Modifications

This PR changes the cache to only watch Numaflow resources to minimize the scope of the cache. The included resources are configurable via ConfigMap.

### Verification

Verified in a local cluster with the following,
```
make start
kubectl apply -k config/samples/
```

### CODE REVIEW

File: internal/sync/cache.go [hunk 1]
```diff
@@ -70,15 +71,15 @@ type ResourceInfo struct {
 }
 
 // LiveStateCache is a cluster caching that stores resource references and ownership
-// references. It also stores custom metadata for resources managed by GitSyncs. It always
-// ensures the cache is up-to-date before returning the resources.
+// references. It also stores custom metadata for includedResources managed by GitSyncs. It always
+// ensures the cache is up-to-date before returning the includedResources.
```
Comment (juliev0) [inline internal/sync/cache.go:75]:
```suggestion
// references. It also stores custom metadata for included resources managed by GitSyncs. It always
// ensures the cache is up-to-date before returning the included resources.
```

Comment (juliev0) [inline internal/sync/cache.go:75]: was this changed on purpose or by accident?

File: internal/sync/cache.go [hunk 2]
```diff
@@ -322,19 +323,59 @@ func (c *liveStateCache) getSyncedCluster() (clustercache.ClusterCache, error) {
...
 		ResourcesFilter: resourcesFilter,
 	}
 
-	return &cacheSettings{clusterSettings, ignoreResourceUpdatesEnabled}
+	return &cacheSettings{clusterSettings, ignoreResourceUpdatesEnabled}, nil
+}
+
+// parseResourceFilter parse the given rules to generate the
+// included resource to be watched during caching.
+func parseResourceFilter(rules string) (*[]ResourceType, error) {
```
[the remark this answers is not in the document]
Comment (juliev0) [reply internal/sync/cache.go:357]: Also, do you feel confident enough about all of this logic to not need a unit test? (If you feel that your manual test covered all the bases, I'm okay with that)

File: internal/sync/cache.go [hunk 3]
```diff
@@ -363,11 +404,26 @@ func NewNoopNormalizer() diff.Normalizer {
 	return &NoopNormalizer{}
 }
 
-// NoopResourceFilter a noop resource filter
-// TODO: add more meaningful resource filter
-type NoopResourceFilter struct {
+// ResourceFilter resource filter based on what are the includedResources included
```
Comment (juliev0) [inline internal/sync/cache.go:407]:
```suggestion
// ResourceFilter filter resources based on allowed Resource Types
```

File: internal/sync/cache.go (hunk 1 above)
Comment (xdevxy) [reply internal/sync/cache.go:75]: yeah this is changed by IDE automatically, will revert it.

File: internal/sync/cache.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (xdevxy) [reply internal/sync/cache.go:357]: I was actually thinking about a unit test, will add it.

File: internal/sync/cache.go [hunk 4]
```diff
@@ -322,19 +323,59 @@ func (c *liveStateCache) getSyncedCluster() (clustercache.ClusterCache, error) {
...
+	}
 	c.cacheSettings = *setting
+	return nil
 }
 
-func (c *liveStateCache) loadCacheSettings() *cacheSettings {
+func (c *liveStateCache) loadCacheSettings() (*cacheSettings, error) {
 	ignoreResourceUpdatesEnabled := false
-	resourcesFilter := &NoopResourceFilter{}
+	globalConfig, err := controllerConfig.GetConfigManagerInstance().GetConfig()
```
Comment (juliev0) [inline internal/sync/cache.go:337]: I guess this is not a field that we will allow to dynamically change without needing to restart the Controller?

Comment (juliev0) [reply internal/sync/cache.go:337]: (I'm okay if we don't deal with that here but we may want to have an Issue for that kind of thing, especially if there are other fields like that.)

Comment (xdevxy) [reply internal/sync/cache.go:337]: yeah, right. Dynamic change shouldn't be allowed.

File: internal/sync/cache.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (juliev0) [reply internal/sync/cache.go:357]: thank you!

File: internal/sync/cache.go (hunk 4 above)
Comment (juliev0) [reply internal/sync/cache.go:337]: Do you figure there would be issues if we enabled that?

Comment (xdevxy) [reply internal/sync/cache.go:337]: no, I don't see any issue as this config is only read during cache init. So I think we should be fine.

File: internal/sync/cache.go [hunk 5]
```diff
@@ -165,7 +166,7 @@ func (c *liveStateCache) getCluster() clustercache.ClusterCache {
 		}
 
 		// TODO: when switch to change event triggering for sync tasking, notify
-		//       the GitSync managing the resources that there are changes happen
+		//       the GitSync managing the includedResources that there are changes happen
```
Comment (juliev0) [inline internal/sync/cache.go:169]:
```suggestion
		//       the GitSync managing the resources that there are changes happen
```

File: internal/sync/cache_test.go [hunk 6]
```diff
@@ -199,3 +199,62 @@ metadata:
...
+			resources, err := parseResourceFilter(tc.rules)
+			if tc.hasErr {
+				assert.NotNil(t, err)
+				assert.Nil(t, resources)
+			} else {
+				assert.Nil(t, err)
+				assert.Equal(t, tc.expectedResources, *resources)
+			}
+		})
+	}
```
Comment (juliev0) [inline internal/sync/cache_test.go:259]: thanks!

File: internal/sync/syncer.go [hunk 7]
```diff
@@ -112,7 +112,7 @@ func (s *Syncer) StopWatching(key string) {
 }
 
 // Start function starts the synchronizer worker group.
-// Each worker keeps picking up tasks (which contains GitSync keys) to sync the resources.
+// Each worker keeps picking up tasks (which contains GitSync keys) to sync the includedResources.
```
Comment (juliev0) [inline internal/sync/syncer.go:115]:
```suggestion
// Each worker keeps picking up tasks (which contains GitSync keys) to sync the resources.
```

File: internal/sync/diff.go [hunk 8]
```diff
@@ -136,7 +136,7 @@ func NewIgnoreNormalizer(overrides map[string]ResourceOverride) (diff.Normalizer
 	return &ignoreNormalizer{patches: patches}, nil
 }
 
-// Normalize removes fields from supplied resource using json paths from matching items of specified resources ignored differences list
+// Normalize removes fields from supplied resource using json paths from matching items of specified includedResources ignored differences list
```
Comment (juliev0) [inline internal/sync/diff.go:139]:
```suggestion
// Normalize removes fields from supplied resource using json paths from matching items of specified resources ignored differences list
```

File: internal/sync/syncer.go [hunk 9]
```diff
@@ -267,7 +270,7 @@ func (r *resourceInfoProviderStub) IsNamespaced(_ schema.GroupKind) (bool, error
 	return false, nil
 }
 
-// sync compares the live state of the watched resources to target state defined in git
+// sync compares the live state of the watched includedResources to target state defined in git
```
Comment (juliev0) [inline internal/sync/syncer.go:273]:
```suggestion
// sync compares the live state of the watched resources to target state defined in git
```

Comment (xdevxy) [reply internal/sync/syncer.go:273]: Oops sorry for these unintended changes. Thanks for catching it!

**Review by juliev0 — APPROVED** [review]

**Pull request merged**